cpp/src/arrow/util/decimal.cc

```diff
@@ -29,8 +29,8 @@ ARROW_EXPORT Status FromString(
   }
 
   int8_t sign = 1;
-  auto charp = s.cbegin();
-  auto end = s.cend();
+  auto charp = s.c_str();
+  auto end = charp + s.length();
```

Comment:
>  I'd like to check to make sure that my changes address this issue.
likely no, there are several places where "end of iterator check" should be added.
for example `char first_char = *charp;`
I believe working with null-terminated c-string is more suitable here.